docs/developer/best-practices/navigation.asciidoc

```diff
@@ -0,0 +1,203 @@
+[[kibana-navigation]]
+== Routing, Navigation and URL
+
+{kib} platform provides a set of tools to help developers built consistent experience around routing and browser navigation.
+Some of that tooling is inside `core`, some is available as part of various plugins.  
+
+The purpose of this guide is to give you (a {kib} contributor}) a high-level overview of available tools and to explain common approaches for handling routing,
+and browser navigation.
+
+This guide covers following topics:
+
+* <<navigating-between-kibana-apps>>
+* <<deep-linking>>
+* <<routing>>
+* <<history-and-location>>
+* <<state-sync>>
+* <<preserve-state>>
+
+
+[[navigating-between-kibana-apps]]
+=== Navigating between {kib} apps
+
+{kib} is a single page application and there is a set of simple rules developers should follow
+to make sure there is no page reload when navigating from one place in {kib} to another. 
+
+**For example**, navigation using native browser apis:  
+
+[source,js]
+----
+window.location.href = core.http.basePath.prepend(`/dashboard/my-dashboard`); // (try to avoid this)
```

Comment:
For some reason this code block has almost no syntax highlighting:

![image](https://user-images.githubusercontent.com/9773803/92716899-b9a0fe80-f35f-11ea-9070-9ea5bbe1ace7.png)

I wonder if it is how our website is set up or does it have to do with `[source,js]`.

Reply:
Great example, but I was initially confused because you say "using native browser apis" but part of the example is using core apis. 

What if we had the "Deep linking into Kibana Apps" section first, so the user is introduced to the core api `core.http.basePath.prepend` and what it does and that it should only be used if there isn't a url generator available?

Then you can follow up with this section and I think the user may understand that you are focusing on the `window.location.href` part of the line. 

Could then even write it as:

```ts
const discoverUrl = discoverUrlGenerator.createUrl({filters, timeRange});
// Avoid this, as it will cause a full page reload.
window.location.href = discoverUrl;
```

Which focuses in on which is the part to avoid.